syntax = "proto3";

service KeyValueStore {
    rpc Put (PutRequest) returns (Response);
    rpc Get (GetRequest) returns (GetResponse);
    rpc Delete (DeleteRequest) returns (Response);
    rpc Update (UpdateRequest) returns (Response);

    rpc Replicate (PutRequest) returns (Response);  // Đảm bảo có phương thức này
    rpc DeleteReplication (DeleteRequest) returns (Response);  // Và phương thức này

    rpc Heartbeat (HeartbeatRequest) returns (HeartbeatResponse);

    rpc SyncData (SyncRequest) returns (SyncResponse);
}

message PutRequest {
    string key = 1;
    string value = 2;
}

message GetRequest {
    string key = 1;
}

message DeleteRequest {
    string key = 1;
}

message UpdateRequest {
    string key = 1;
    string new_value = 2;
}

message Response {
    bool success = 1;
    string message = 2;
}

message GetResponse {
    bool success = 1;
    string value = 2;
}

message HeartbeatRequest {}

message HeartbeatResponse {
    bool success = 1;
}

message SyncRequest {}

message SyncResponse {
    map<string, string> data = 1; // Trả về toàn bộ dữ liệu
}